vendor/bat-native-ledger/src/bat/ledger/internal/database/database_activity_info.cc

```diff
@@ -0,0 +1,676 @@
+        "UNIQUE (publisher_id, month, year, reconcile_stamp) "
+        "CONSTRAINT fk_%s_publisher_id"
+        "    FOREIGN KEY (publisher_id)"
+        "    REFERENCES publisher_info (publisher_id)"
+        "    ON DELETE CASCADE"
+      ")",
+      table_name_,
+      table_name_);
+
+  auto command = ledger::DBCommand::New();
+  command->type = ledger::DBCommand::Type::EXECUTE;
+  command->command = query;
+  transaction->commands.push_back(std::move(command));
+
+  return true;
+}
+
+bool DatabaseActivityInfo::CreateTableV6(ledger::DBTransaction* transaction) {
+  const std::string query = base::StringPrintf(
+      "CREATE TABLE %s ("
+        "publisher_id LONGVARCHAR NOT NULL,"
+        "duration INTEGER DEFAULT 0 NOT NULL,"
+        "visits INTEGER DEFAULT 0 NOT NULL,"
+        "score DOUBLE DEFAULT 0 NOT NULL,"
+        "percent INTEGER DEFAULT 0 NOT NULL,"
+        "weight DOUBLE DEFAULT 0 NOT NULL,"
+        "reconcile_stamp INTEGER DEFAULT 0 NOT NULL,"
+        "CONSTRAINT activity_unique "
+        "UNIQUE (publisher_id, reconcile_stamp) "
+        "CONSTRAINT fk_%s_publisher_id"
+        "    FOREIGN KEY (publisher_id)"
+        "    REFERENCES publisher_info (publisher_id)"
+        "    ON DELETE CASCADE"
+      ")",
+      table_name_,
+      table_name_);
+
+  auto command = ledger::DBCommand::New();
+  command->type = ledger::DBCommand::Type::EXECUTE;
+  command->command = query;
+  transaction->commands.push_back(std::move(command));
+
+  return true;
+}
+
+bool DatabaseActivityInfo::CreateTableV15(ledger::DBTransaction* transaction) {
+  const std::string query = base::StringPrintf(
+      "CREATE TABLE %s ("
+        "publisher_id LONGVARCHAR NOT NULL,"
+        "duration INTEGER DEFAULT 0 NOT NULL,"
+        "visits INTEGER DEFAULT 0 NOT NULL,"
+        "score DOUBLE DEFAULT 0 NOT NULL,"
+        "percent INTEGER DEFAULT 0 NOT NULL,"
+        "weight DOUBLE DEFAULT 0 NOT NULL,"
+        "reconcile_stamp INTEGER DEFAULT 0 NOT NULL,"
+        "CONSTRAINT activity_unique "
+        "UNIQUE (publisher_id, reconcile_stamp)"
+      ")",
+      table_name_);
+
+  auto command = ledger::DBCommand::New();
+  command->type = ledger::DBCommand::Type::EXECUTE;
+  command->command = query;
+  transaction->commands.push_back(std::move(command));
+
+  return true;
+}
+
+bool DatabaseActivityInfo::CreateIndexV2(ledger::DBTransaction* transaction) {
+  return this->InsertIndex(transaction, table_name_, "publisher_id");
+}
+
+bool DatabaseActivityInfo::CreateIndexV4(ledger::DBTransaction* transaction) {
+  return this->InsertIndex(transaction, table_name_, "publisher_id");
+}
+
+bool DatabaseActivityInfo::CreateIndexV6(ledger::DBTransaction* transaction) {
+  return this->InsertIndex(transaction, table_name_, "publisher_id");
+}
+
+bool DatabaseActivityInfo::CreateIndexV15(ledger::DBTransaction* transaction) {
+  return this->InsertIndex(transaction, table_name_, "publisher_id");
+}
+
+bool DatabaseActivityInfo::Migrate(
```

Comment:
I'd propose that all the DB Migration code would be better off in a dedicated Migration oriented object, rather than in this multi-purpose one.

Reply:
we have Migration file where we handle basic logic, then we have per table specific migration in table specific file

Reply:
Ok. Personally, I'd still go for 'db_table_migration.cc' - which can then becomes unit testable.

Reply:
this approach can be fully unit testable afaik